### PR TITLE
[20.09] nginx: Fix off-by-one in DNS resolver heap write

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

Fixes CVE-2021-23027

https://www.openwall.com/lists/oss-security/2021/05/25/5

I originally tried applying https://nginx.org/download/patch.2021.resolver.txt, but that patch for some reason didn't find the src file.

```sh
❯ nix-build -A nginx
these derivations will be built:
  /nix/store/vf7sgcpb59jkk60l8aykkmsb1z9aijii-nginx-1.18.0.drv
building '/nix/store/vf7sgcpb59jkk60l8aykkmsb1z9aijii-nginx-1.18.0.drv' on 'ssh://hexa@phoibe.cysec.de'...
unpacking sources
unpacking source archive /nix/store/hxa3gqx34w019jgdjp756pdfwnlq5nry-nginx-1.18.0.tar.gz
source root is nginx-1.18.0
setting SOURCE_DATE_EPOCH to timestamp 1587478146 of file nginx-1.18.0/CHANGES
patching sources
# prePatch ls works!
src/core/ngx_resolver.c
applying patch /nix/store/sc131zdz663kid2crlba7axarykyjirx-nix-etag-1.15.4.patch
patching file src/http/ngx_http_core_module.c
Hunk #1 succeeded at 1657 (offset 74 lines).
Hunk #2 succeeded at 1674 (offset 74 lines).
applying patch /nix/store/zwz37yd9whxjj8inym5vg21b5plp1k0x-nix-skip-check-logs-path.patch
patching file auto/install
applying patch /nix/store/6ib35bgpv3c943csgwhhckzfl3pnr4za-CVE-2021-23017.patch
can't find file to patch at input line 3
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|--- src/core/ngx_resolver.c
|+++ src/core/ngx_resolver.c
--------------------------
File to patch:
Skip this patch? [y]
Skipping patch.
1 out of 1 hunk ignored

```

Related: #124400 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/http/nginx/CVE-2021-23017.patch [hunk 1]
```diff
@@ -0,0 +1,24 @@
+diff --git a/src/core/ngx_resolver.c b/src/core/ngx_resolver.c
```
Comment (andir) [inline pkgs/servers/http/nginx/CVE-2021-23017.patch:1]: Can you add the reference to the source of the patch as comment at the top of this file, please?

**Review by andir — CHANGES_REQUESTED** [review]

**Comment by lukegb** [comment]
Hmm, `fetchpatch` should work, but you might need to set `extraPrefix = '';` since it's a raw un-prefixed patch

**Comment by mweinelt** [comment]
> Hmm, `fetchpatch` should work, but you might need to set `extraPrefix = '';` since it's a raw un-prefixed patch

Woah black magic. Thank you! Updating in a minute.

File: pkgs/servers/http/nginx/CVE-2021-23017.patch (hunk 1 above)
Comment (mweinelt) [reply pkgs/servers/http/nginx/CVE-2021-23017.patch]: Obsolete.

File: pkgs/servers/http/nginx/generic.nix [hunk 2]
```diff
@@ -116,6 +116,14 @@ stdenv.mkDerivation {
       '';
     })
     ./nix-skip-check-logs-path.patch
+    (fetchpatch {
```
Comment (andir) [inline pkgs/servers/http/nginx/generic.nix:119]: Why bother with fetchpatch at all? The file is a plain text file that just applies as patch. It doesn't matter much but `fetchpatch` shouldn't be required in this case.

Comment (mweinelt) [reply pkgs/servers/http/nginx/generic.nix:119]: Because it needs the `extraPrefix` special casing, that is only in `fetchpatch`.

Comment (andir) [reply pkgs/servers/http/nginx/generic.nix:119]: Oh! D'oh!

**Review by ajs124 — APPROVED** [review]

File: pkgs/servers/http/nginx/generic.nix (hunk 2 above)
Comment (risicle) [reply pkgs/servers/http/nginx/generic.nix:119]: Isn't it needed to prefix the paths in this case?

Comment (mweinelt) [reply pkgs/servers/http/nginx/generic.nix:119]: Yep.

Comment (risicle) [reply pkgs/servers/http/nginx/generic.nix:119]: :roll_eyes: github being slow to sync today...

**Comment by risicle** [comment]
`nixpkgs-review` happy, macos 10.15 apart from `devpi-server` and `devpi-client` which hate the sandbox on darwin (#124428)

`nixosTests.nginx` `nixosTests.nginx-etag` `nixosTests.nginx-pubhtml` `nixosTests.nginx-sandbox` `nixosTests.php.fpm`happy, nixos x86_64

**Pull request merged**